File: web/src/api/types/training.ts
import type { JobTargetRef } from './job-target';
import type { ProjectProfile } from './project';
import type { UserProfile } from './profile';
import type { PromptSetSummary } from './prompt';

export interface WeaknessTag {
  id: string;
  kind: string;
  label: string;
  severity: number;
  frequency: number;
  last_seen_at: string;
  evidence_session_id: string;
}

export interface TrainingEvaluation {
  score: number;
  score_breakdown: Record<string, number>;
  headline?: string;
  strengths: string[];
  gaps: string[];
  suggestion?: string;
  followup_intent?: string;
  followup_question?: string;
  followup_expected_points?: string[];
}

export interface TrainingTurn {
  id: string;
  turn_index: number;
  question: string;
  expected_points: string[];
  answer: string;
  evaluation?: TrainingEvaluation;
}

export interface TrainingSession {
  id: string;
  mode: 'basics' | 'project';
  topic?: string;
  project_id?: string;
  job_target_id?: string;
  job_target_analysis_id?: string;
  prompt_set_id?: string;
  intensity: string;
  status: string;
  max_turns: number;
  total_score: number;
  review_id?: string;
  turns: TrainingTurn[];
  project?: ProjectProfile;
  job_target?: JobTargetRef | null;
  prompt_set?: PromptSetSummary | null;
}

export interface TrainingSessionSummary {
  id: string;
  mode: string;
  topic?: string;
  project_name?: string;
  status: string;
  total_score: number;
  review_id?: string;
  updated_at: string;
  job_target?: JobTargetRef | null;
  prompt_set_id?: string;
  prompt_set?: PromptSetSummary | null;
}

export interface RuntimeTraceEntry {
  flow: string;
  phase: string;
  status: string;
  code?: string;
  message?: string;
  attempt?: number;
  tool_name?: string;
}

export interface RuntimeTrace {
  entries: RuntimeTraceEntry[];
}

export interface MemoryRetrievalHit {
  source: string;
  memory_index_id?: string;
  ref_table?: string;
  ref_id?: string;
  scope_type?: string;
  scope_id?: string;
  topic?: string;
  summary?: string;
  rule_score?: number;
  vector_score?: number;
  rerank_score?: number;
  final_score?: number;
  reason?: string;
}

export interface MemoryRetrievalTrace {
  memory_type: string;
  query?: string;
  strategy: string;
  candidate_count: number;
  selected_count: number;
  fallback_used?: boolean;
  fallback_reason?: string;
  hits: MemoryRetrievalHit[];
}

export interface RetrievalTrace {
  generated_at: string;
  topic?: string;
  project_id?: string;
  job_target_id?: string;
  observations?: MemoryRetrievalTrace | null;
  session_summaries?: MemoryRetrievalTrace | null;
}

export interface ReviewCard {
  id: string;
  session_id: string;
  job_target_id?: string;
  job_target_analysis_id?: string;
  prompt_set_id?: string;
  overall: string;
  top_fix?: string;
  top_fix_reason?: string;
  highlights: string[];
  gaps: string[];
  suggested_topics: string[];
  next_training_focus: string[];
  recommended_next?: {
    mode: 'basics' | 'project';
    topic?: string;
    project_id?: string;
    reason?: string;
  } | null;
  score_breakdown: Record<string, number>;
  retrieval_trace?: RetrievalTrace | null;
  job_target?: JobTargetRef | null;
  prompt_set?: PromptSetSummary | null;
}

export interface EvaluationLogEntry {
  id: number;
  session_id: string;
  turn_id?: string;
  flow_name: string;
  model_name?: string;
  prompt_set_id?: string;
  prompt_hash?: string;
  raw_output?: string;
  runtime_trace?: RuntimeTrace | null;
  latency_ms: number;
  created_at: string;
}

export interface Dashboard {
  profile: UserProfile | null;
  weaknesses: WeaknessTag[];
  recent_sessions: TrainingSessionSummary[];
  current_session?: TrainingSessionSummary | null;
  today_focus: string;
  recommended_track: string;
  active_job_target?: JobTargetRef | null;
  recommendation_scope: 'generic' | 'job_target';
  days_until_deadline?: number;
}

export interface StreamEvent {
  type:
    | 'phase'
    | 'context'
    | 'trace'
    | 'reasoning'
    | 'content'
    | 'status'
    | 'result'
    | 'error';
  code?: string;
  phase?: string;
  name?: string;
  text?: string;
  message?: string;
  data?: unknown;
}

export interface PaginatedList<T> {
  items: T[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface WeaknessTrendPoint {
  session_id: string;
  severity: number;
  created_at: string;
}

export interface WeaknessTrend {
  id: string;
  kind: string;
  label: string;
  points: WeaknessTrendPoint[];
}

export interface ReviewScheduleItem {
  id: number;
  session_id: string;
  review_card_id?: string;
  weakness_tag_id?: string;
  weakness_kind?: string;
  weakness_label?: string;
  topic?: string;
  next_review_at: string;
  interval_days: number;
}

File: server/internal/domain/training.go
package domain

import "time"

type TrainingSession struct {
	ID                  string            `json:"id"`
	Mode                string            `json:"mode"`
	Topic               string            `json:"topic,omitempty"`
	ProjectID           string            `json:"project_id,omitempty"`
	JobTargetID         string            `json:"job_target_id,omitempty"`
	JobTargetAnalysisID string            `json:"job_target_analysis_id,omitempty"`
	PromptSetID         string            `json:"prompt_set_id,omitempty"`
	Intensity           string            `json:"intensity"`
	Status              string            `json:"status"`
	MaxTurns            int               `json:"max_turns"`
	TotalScore          float64           `json:"total_score"`
	StartedAt           *time.Time        `json:"started_at,omitempty"`
	EndedAt             *time.Time        `json:"ended_at,omitempty"`
	CreatedAt           time.Time         `json:"created_at"`
	UpdatedAt           time.Time         `json:"updated_at"`
	ReviewID            string            `json:"review_id,omitempty"`
	Turns               []TrainingTurn    `json:"turns,omitempty"`
	Project             *ProjectProfile   `json:"project,omitempty"`
	JobTarget           *JobTargetRef     `json:"job_target,omitempty"`
	PromptSet           *PromptSetSummary `json:"prompt_set,omitempty"`
}

type TrainingTurn struct {
	ID             string            `json:"id"`
	SessionID      string            `json:"session_id"`
	TurnIndex      int               `json:"turn_index"`
	Stage          string            `json:"stage"`
	Question       string            `json:"question"`
	ExpectedPoints []string          `json:"expected_points"`
	Answer         string            `json:"answer"`
	Evaluation     *EvaluationResult `json:"evaluation,omitempty"`
	WeaknessHits   []WeaknessHit     `json:"weakness_hits,omitempty"`
	CreatedAt      time.Time         `json:"created_at"`
	UpdatedAt      time.Time         `json:"updated_at"`
}

type EvaluationResult struct {
	Score            float64            `json:"score"`
	ScoreBreakdown   map[string]float64 `json:"score_breakdown"`
	Headline         string             `json:"headline,omitempty"`
	Strengths        []string           `json:"strengths"`
	Gaps             []string           `json:"gaps"`
	Suggestion       string             `json:"suggestion,omitempty"`
	FollowupIntent   string             `json:"followup_intent,omitempty"`
	FollowupQuestion string             `json:"followup_question,omitempty"`
	FollowupPoints   []string           `json:"followup_expected_points,omitempty"`
	WeaknessHits     []WeaknessHit      `json:"weakness_hits"`
}

type WeaknessHit struct {
	Kind     string  `json:"kind"`
	Label    string  `json:"label"`
	Severity float64 `json:"severity"`
}

type WeaknessTag struct {
	ID                string    `json:"id"`
	Kind              string    `json:"kind"`
	Label             string    `json:"label"`
	Severity          float64   `json:"severity"`
	Frequency         int       `json:"frequency"`
	LastSeenAt        time.Time `json:"last_seen_at"`
	EvidenceSessionID string    `json:"evidence_session_id"`
}

type WeaknessTrendPoint struct {
	SessionID string  `json:"session_id"`
	Severity  float64 `json:"severity"`
	CreatedAt string  `json:"created_at"`
}

type WeaknessTrend struct {
	ID     string               `json:"id"`
	Kind   string               `json:"kind"`
	Label  string               `json:"label"`
	Points []WeaknessTrendPoint `json:"points"`
}

type ReviewScheduleItem struct {
	ID            int64     `json:"id"`
	SessionID     string    `json:"session_id"`
	ReviewCardID  string    `json:"review_card_id,omitempty"`
	WeaknessTagID string    `json:"weakness_tag_id,omitempty"`
	WeaknessKind  string    `json:"weakness_kind,omitempty"`
	WeaknessLabel string    `json:"weakness_label,omitempty"`
	Topic         string    `json:"topic,omitempty"`
	NextReviewAt  time.Time `json:"next_review_at"`
	IntervalDays  int       `json:"interval_days"`
	EaseFactor    float64   `json:"ease_factor"`
	CreatedAt     time.Time `json:"created_at"`
}

type ReviewCard struct {
	ID                  string             `json:"id"`
	SessionID           string             `json:"session_id"`
	JobTargetID         string             `json:"job_target_id,omitempty"`
	JobTargetAnalysisID string             `json:"job_target_analysis_id,omitempty"`
	PromptSetID         string             `json:"prompt_set_id,omitempty"`
	Overall             string             `json:"overall"`
	TopFix              string             `json:"top_fix,omitempty"`
	TopFixReason        string             `json:"top_fix_reason,omitempty"`
	Highlights          []string           `json:"highlights"`
	Gaps                []string           `json:"gaps"`
	SuggestedTopics     []string           `json:"suggested_topics"`
	NextTrainingFocus   []string           `json:"next_training_focus"`
	RecommendedNext     *NextSession       `json:"recommended_next,omitempty"`
	RetrievalTrace      *RetrievalTrace    `json:"retrieval_trace,omitempty"`
	ScoreBreakdown      map[string]float64 `json:"score_breakdown"`
	CreatedAt           time.Time          `json:"created_at"`
	JobTarget           *JobTargetRef      `json:"job_target,omitempty"`
	PromptSet           *PromptSetSummary  `json:"prompt_set,omitempty"`
}

type NextSession struct {
	Mode      string `json:"mode"`
	Topic     string `json:"topic,omitempty"`
	ProjectID string `json:"project_id,omitempty"`
	Reason    string `json:"reason,omitempty"`
}

type TrainingSessionSummary struct {
	ID          string            `json:"id"`
	Mode        string            `json:"mode"`
	Topic       string            `json:"topic,omitempty"`
	ProjectName string            `json:"project_name,omitempty"`
	Status      string            `json:"status"`
	TotalScore  float64           `json:"total_score"`
	ReviewID    string            `json:"review_id,omitempty"`
	UpdatedAt   time.Time         `json:"updated_at"`
	JobTarget   *JobTargetRef     `json:"job_target,omitempty"`
	PromptSetID string            `json:"prompt_set_id,omitempty"`
	PromptSet   *PromptSetSummary `json:"prompt_set,omitempty"`
}

type ListSessionsRequest struct {
	Page    int    `form:"page"`
	PerPage int    `form:"per_page"`
	Mode    string `form:"mode"`
	Topic   string `form:"topic"`
	Status  string `form:"status"`
}

type PaginatedList[T any] struct {
	Items      []T `json:"items"`
	Total      int `json:"total"`
	Page       int `json:"page"`
	PerPage    int `json:"per_page"`
	TotalPages int `json:"total_pages"`
}

type CreateSessionRequest struct {
	Mode                  string `json:"mode" binding:"required"`
	Topic                 string `json:"topic"`
	ProjectID             string `json:"project_id"`
	JobTargetID           string `json:"job_target_id"`
	PromptSetID           string `json:"prompt_set_id"`
	IgnoreActiveJobTarget bool   `json:"ignore_active_job_target,omitempty"`
	Intensity             string `json:"intensity" binding:"required"`
	MaxTurns              int    `json:"max_turns,omitempty"`
}

type ExportSessionsRequest struct {
	SessionIDs []string `json:"session_ids" binding:"required"`
	Format     string   `json:"format" binding:"required"`
}

type SubmitAnswerRequest struct {
	Answer string `json:"answer" binding:"required"`
}

type StreamEvent struct {
	Type    string `json:"type"`
	Code    string `json:"code,omitempty"`
	Phase   string `json:"phase,omitempty"`
	Name    string `json:"name,omitempty"`
	Text    string `json:"text,omitempty"`
	Message string `json:"message,omitempty"`
	Data    any    `json:"data,omitempty"`
}
